library/pkcs7_write.c

```diff
@@ -0,0 +1,829 @@
+ *							OID (hash Alg)						->^
+ *						CONSTRUCTED | SEQUENCE 					->^
+ *							OID (Signature Alg (RSA)) 			->^
+ *						OCTET STRING (signature) 				->pkcs7_set_signature
+ * }
+ */
+
+//struct for containing information needed to generate the pkcs7
+//used for less function arguments
+typedef struct mbedtls_pkcs7_info {
+    unsigned char **crts;  // signing crt DER
+    size_t *crt_sizes;
+    unsigned char **keys;  // signing key DER
+    size_t *key_sizes;
+    int key_pairs;
+    const unsigned char *data;
+    int data_size;
+    mbedtls_md_type_t hash_funct;
+    const char *hash_funct_oid;
+
+} mbedtls_pkcs7_info;
+
+/*
+ *Taken from MBEDTLS, mbedtls-mbedtls2.23.0/programs/util/pem2der.c many thanks
+ *to these great folks Some things were changed though, like memory allocation
+ *THIS ALLOCATES MEMORY, FREE SOMETIME AFTER CALLING
+ */
+int mbedtls_convert_pem_to_der(const unsigned char *input, size_t ilen,
+                       unsigned char **output, size_t *olen)
+{
+    int ret;
+    const unsigned char *s1, *s2, *end = input + ilen;
+    size_t len = 0;
+
+    s1 = (unsigned char *) strstr( (const char *) input, "-----BEGIN" );
+    if( s1 == NULL ) return (-1);
+
+    s2 = (unsigned char *) strstr( (const char *) input, "-----END" );
+    if( s2 == NULL ) return (-1);
+
+    s1 += 10;
+    while( s1 < end && *s1 != '-') 
+        s1++;
+    while( s1 < end && *s1 == '-') 
+        s1++;
+    if( *s1 == '\r') 
+        s1++;
+    if( *s1 == '\n') 
+        s1++;
+
+    if( s2 <= s1 || s2 > end) 
+    	return ( -1 );
+
+    ret = mbedtls_base64_decode( NULL, 0, &len, (const unsigned char *) s1, s2 - s1);
+    if( ret == MBEDTLS_ERR_BASE64_INVALID_CHARACTER ) 
+    {
+        return ( ret );
+    }
+    // free this ouside of function
+    *output = malloc( len );
+    if( !*output ) 
+    {
+        return ( MBEDTLS_ERR_PKCS7_ALLOC_FAILED );
+    }
+    *olen = len;
+
+    if( ( ret = mbedtls_base64_decode( *output, len, &len,
+                                     (const unsigned char *) s1, s2 - s1 ) ) != 0 ) 
+    {
+        return ( ret );
+    }
+
+    return ( 0 );
+}
+
+static size_t get_leading_whitespace( unsigned char *data, size_t data_size )
+{
+    size_t white_spaceSize = 0;
+    while( white_spaceSize < data_size && data[white_spaceSize] == 0x00 )
+        white_spaceSize++;
+    return ( white_spaceSize );
+}
+
+/*
+ * \brief 		given a data buffer it generates the desired hash
+ * \param data, data to be hashed
+ * \param size , length of buff
+ * \param hash_funct, mbedtls_md_type, message digest type
+ * \param outBuff, the resulting hash, NOTE: REMEMBER TO UNALLOC THIS MEMORY
+ * \param outBuffSize, should be alg->size
+ * \return 0 or err number
+ */
+static int pkcs7_hash_create( const unsigned char *data, size_t size, int hash_funct,
+           char **out_hash, size_t *out_hash_size )
+{
+    const mbedtls_md_info_t *md_info;
+    mbedtls_md_context_t ctx;
+    int rc;
```

Comment:
Style: we tend to call that `ret`, and I think using the same name everywhere is better for readability.